### PR TITLE
feat: Add device last connected metrics

### LINKED ISSUE #1470: Device Last Connected
# 🚀 Feature Request

### Relevant Package [**REQUIRED**]

 This feature request is for when Device commands (Get & Set) are executed and when a Device sends Async data

### Description [**REQUIRED**]
Implement Last Connect capability using a `Gauge` metric for each device that the service is managing.

The `gometrics.Gauge` metrics for each device can be stored in the Device Cache here with new access by name functions.
https://github.com/edgexfoundry/device-sdk-go/blob/main/internal/cache/devices.go#L30C1-L33C2
```go
type deviceCache struct {
	deviceMap map[string]*models.Device // key is Device name
        lastConnected map[string] gometrics.Gauge
	mutex     sync.RWMutex
}
```
The metric is added/removed when a device is added or removed. The metric must be registered/unregistered vi the `MetricsManager`
The naming scheme for the metric is `LastConnected-{device-name}` . We may need to deal with any special characters in the name that might impact the JSON when the metric is reported via the MessageBus. This may be an enhancement to metrics.Reporter.

Here are examples from App SDK for storing, initializing metrics for function pipelines:

https://github.com/edgexfoundry/app-functions-sdk-go/blob/main/pkg/interfaces/service.go#L61-L63
https://github.com/edgexfoundry/app-functions-sdk-go/blob/main/internal/runtime/runtime.go#L57
https://github.com/edgexfoundry/app-functions-sdk-go/blob/main/internal/runtime/runtime.go#L183-L199
This uses similar naming scheme. Base metric name with pipeline name appended.

### Describe the solution you'd like
 If you have a solution in mind, please describe it.

### Describe alternatives you've considered
 Have you considered any alternative solutions or workarounds?

### PR DESCRIPTION
**Pull request by lindseysimple**
fixes #1470. Add device last connected metrics to Get/Set command.

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/device-sdk-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [x] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [x] I have opened a PR for the related docs change (if not, why?) - https://github.com/edgexfoundry/edgex-docs/pull/1231
  <link to docs PR>

### CODE REVIEW

**Review by cloudxxx8 — COMMENTED** [review]
Use URL escape for device name if `configuration.Service.EnableNameFieldEscape` is `true`
Reference:
https://github.com/edgexfoundry/device-sdk-go/pull/1502/files
https://github.com/edgexfoundry/device-sdk-go/pull/1506/files

**Review by cloudxxx8 — COMMENTED** [review]
async value should also be handled

**Comment by lindseysimple** [comment]
Change's made for `URL escape for device name` and update `LastConnected` metric for async value.

File: internal/cache/devices.go [hunk 1]
```diff
@@ -25,24 +39,64 @@ type DeviceCache interface {
...
 	return dc
 }
 
+func registerMetric(deviceName string, metric interface{}, dic *di.Container) {
+	metricsManager := bootstrapContainer.MetricsManagerFrom(dic.Get)
+	lc := bootstrapContainer.LoggingClientFrom(dic.Get)
+	configuration := container.ConfigurationFrom(dic.Get)
+	if configuration.Service.EnableNameFieldEscape {
+		deviceName = common.URLEncode(deviceName)
+	}
```
Comment (lenny-intel) [inline internal/cache/devices.go:78]: Not sure we need to be doing this here. Currently there are no restrictions on the metric name other then non-empty string. IMO the escaping needs to be handled in the Reporter here: https://github.com/edgexfoundry/go-mod-bootstrap/blob/main/bootstrap/metrics/reporter.go#L186

File: internal/cache/devices.go [hunk 2]
```diff
@@ -25,24 +39,64 @@ type DeviceCache interface {
...
+		lc.Infof("%s metric has been registered and will be reported (if enabled)", registeredName)
+	}
+}
+
+func unregisterMetric(deviceName string, dic *di.Container) {
+	metricsManager := bootstrapContainer.MetricsManagerFrom(dic.Get)
+	configuration := container.ConfigurationFrom(dic.Get)
+	if configuration.Service.EnableNameFieldEscape {
+		deviceName = common.URLEncode(deviceName)
+	}
```
Comment (lenny-intel) [inline internal/cache/devices.go:94]: See previous comment

File: internal/cache/devices.go [hunk 3]
```diff
@@ -152,3 +216,16 @@ func CheckProfileNotUsed(profileName string) bool {
...
+	defer d.mutex.RUnlock()
+
+	g := d.lastConnected[name]
+	g.Update(time.Now().UnixNano())
+}
+
+func (d *deviceCache) GetLastConnectedByName(name string) int64 {
+	g := d.lastConnected[name]
+	return g.Value()
+}
```
Comment (lenny-intel) [inline internal/cache/devices.go:231]: Need new unit tests for these.

**Review by lenny-intel — CHANGES_REQUESTED** [review]
Nice Job!! Just a few comments.

File: internal/cache/devices.go (hunk 3 above)
Comment (lindseysimple) [reply internal/cache/devices.go]: Unit test added.

File: internal/cache/devices.go (hunk 1 above)
Comment (lindseysimple) [reply internal/cache/devices.go:78]: Let's discuss this with more details on the device WG meeting tomorrow, thanks Lenny.

Comment (lindseysimple) [reply internal/cache/devices.go]: As Lenny indicated in the meeting, the device name only is included in the metric message payload, removed the `EnableNameFieldEscape` check here as the full publish topic name is `edgex/telemetry/device-simple/LastConnected`.

@lenny-intel change's done and is ready for review, thanks!

File: internal/cache/devices.go [hunk 4]
```diff
@@ -72,13 +69,9 @@ func newDeviceCache(devices []models.Device, dic *di.Container) DeviceCache {
...
 	metricsManager := bootstrapContainer.MetricsManagerFrom(dic.Get)
 	lc := bootstrapContainer.LoggingClientFrom(dic.Get)
-	configuration := container.ConfigurationFrom(dic.Get)
-	if configuration.Service.EnableNameFieldEscape {
-		deviceName = common.URLEncode(deviceName)
-	}
 	registeredName := strings.Replace(lastConnectedPrefix, deviceNameText, deviceName, 1)
 
-	err := metricsManager.Register(registeredName, metric, map[string]string{"device": deviceName})
+	err := metricsManager.Register(registeredName, metric, nil)
```
[the remark this answers is not in the document]
Comment (lindseysimple) [reply internal/cache/devices.go]: Done, tag added back. Thanks.

**Review by lenny-intel — APPROVED** [review]
LGTM

**Pull request merged**